flink-core/src/main/java/org/apache/flink/core/plugin/DirectoryBasedPluginDescriptorsFactory.java

```diff
@@ -0,0 +1,108 @@
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to create a collection of {@link PluginDescriptor} based on directory structure for a given plugin
+ * root folder.
+ *
+ * <p>The expected structure is as follows: the given plugins root folder, containing the plugins folder. One plugin folder
+ * contains all resources (jar files) belonging to a plugin. The name of the plugin folder becomes the plugin id.
+ * <pre>
+ * plugins-root-folder/
+ *	         |------------plugin-a/ (folder of plugin a)
+ *	         |	              |-plugin-a-1.jar (the jars containing the classes of plugin a)
+ *	         |                |-plugin-a-2.jar
+ *	         |                |-...
+ *	         |
+ *	         |------------plugin-b/
+ *		     |                |-plugin-b-1.jar
+ *		    ...               |-...
+ * </pre>
+ */
+public class DirectoryBasedPluginDescriptorsFactory {
+
+	private static final Logger log = LoggerFactory.getLogger(DirectoryBasedPluginDescriptorsFactory.class);
+
+	private static final PathMatcher JAR_FILE_MATCHER =
+		FileSystems.getDefault().getPathMatcher("glob:**.jar");
+
+	private final Path pluginsRootDir;
+
+	public DirectoryBasedPluginDescriptorsFactory(Path pluginsRootDir) {
+		this.pluginsRootDir = pluginsRootDir;
+	}
+
+	//TODO this can go into an interface if there would be more strategies in the future.
+	public Collection<PluginDescriptor> createPluginDescriptors() throws IOException {
+		return Files.list(pluginsRootDir)
+			.filter((Path path) -> Files.isDirectory(path))
+			.map(FunctionUtils.uncheckedFunction(this::createPluginDescriptorForSubDirectory))
+			.filter(Optional::isPresent)
+			.map(Optional::get)
+			.collect(Collectors.toList());
+	}
+
+	private Optional<PluginDescriptor> createPluginDescriptorForSubDirectory(Path subDirectory) {
+		Optional<URL[]> jarURLsFromDirectory = createJarURLsFromDirectory(subDirectory);
+		if (jarURLsFromDirectory.isPresent()) {
+			URL[] urls = jarURLsFromDirectory.get();
+			// we sort the urls for the benefit of having a stable/reproducible order of jars.
+			Arrays.sort(urls, Comparator.comparing(URL::toString));
+			//TODO: This class could be extended to parse exclude-pattern from a optional text files in the plugin directories.
+			return Optional.of(
+				new PluginDescriptor(
+					subDirectory.getFileName().toString(),
+					urls,
+					new String[0]));
+		} else {
+			return Optional.empty();
+		}
+	}
+
+	private static Optional<URL[]> createJarURLsFromDirectory(Path subDirectory) {
+		URL[] urls = null;
+		try {
+			urls = Files.list(subDirectory)
+				.filter((Path p) -> Files.isRegularFile(p) && JAR_FILE_MATCHER.matches(p))
+				.map(FunctionUtils.uncheckedFunction((Path p) -> p.toUri().toURL()))
+				.toArray(URL[]::new);
+		} catch (Exception ex) {
+			log.warn("Exception while collecting the URLs of jar-files in plugin sub-directory {}.", subDirectory, ex);
```

Comment:
The reason was because we did not define if failures for individual plugins should affect execution. We can make this more strict and say that all plugin failures must be handled by the user first.